language/compiler/ir-to-bytecode/syntax/src/lexer.rs

```diff
@@ -311,41 +132,248 @@ impl<'input, 'builder> Lexer<'input, 'builder> {
         let text = self.text.trim_start();
         let whitespace = self.text.len() - text.len();
         let start_offset = self.consumed + whitespace;
-        if text.is_empty() {
-            self.text = text;
-            self.consumed = start_offset;
-            self.token = (start_offset, Token(Tok::EOF, ""), start_offset);
-            Ok(())
-        } else {
-            let mut longest_match = 0;
-            let mut index: Option<usize> = None;
-            for (i, token_exp) in self.regex_vec.iter().enumerate() {
-                if let Some(m) = token_exp.find(text) {
-                    let len = m.end();
-                    if len >= longest_match {
-                        longest_match = len;
-                        index = Some(i.try_into().unwrap());
+        let c: char = match text.chars().next() {
+            Some(next_char) => next_char,
+            None => {
+                self.text = text;
+                self.consumed = start_offset;
+                self.token = (start_offset, Token(Tok::EOF, ""), start_offset);
+                return Ok(());
+            }
+        };
+        let (tok, len) = match c {
+            '0'..='9' => {
+                if (text.starts_with("0x") || text.starts_with("0X")) && text.len() > 2 {
+                    let hex_len = get_hex_digits_len(&text[2..]);
+                    if hex_len == 0 {
+                        // Fall back to treating this as a "0" token.
+                        (Tok::U64Value, 1)
+                    } else {
+                        (Tok::AccountAddressValue, 2 + hex_len)
+                    }
+                } else {
+                    (Tok::U64Value, get_decimal_digits_len(&text))
+                }
+            }
+            'a'..='z' | 'A'..='Z' | '$' | '_' => {
```

Comment:
What's this `$` here for? Do we have `$` in identifiers? @runtian-zhou @sblackshear

Reply:
The old lalrpop file allowed them. E.G.:
Name: String = <s:r"[a-zA-Z$_][a-zA-Z0-9$_]*">